pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -27,13 +27,64 @@ namespace {
 using namespace Pennylane::Util;
 using Kokkos::kokkos_swap;
 using Pennylane::Gates::GateOperation;
+using Pennylane::LightningKokkos::Util::ControlBitPatterns;
+using Pennylane::LightningKokkos::Util::generateBitPatterns;
+using Pennylane::LightningKokkos::Util::parity_2_offset;
+using Pennylane::LightningKokkos::Util::reverseWires;
 using Pennylane::LightningKokkos::Util::vector2view;
 } // namespace
 /// @endcond
 
 namespace Pennylane::LightningKokkos::Functors {
 
-template <class PrecisionT, class FuncT> class applyNC1Functor {
+template <class PrecisionT, class FuncT, bool has_controls>
+class applyNC1Functor {};
+
+template <class PrecisionT, class FuncT>
+class applyNC1Functor<PrecisionT, FuncT, true> {
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+
+  public:
+    template <class ExecutionSpace>
+    applyNC1Functor([[maybe_unused]] ExecutionSpace exec,
+                    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                    std::size_t num_qubits,
+                    const std::vector<std::size_t> &controlled_wires,
+                    const std::vector<bool> &controlled_values,
+                    const std::vector<std::size_t> &wires, FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+
+        std::tie(parity, rev_wires) =
```

Comment:
Couldn't we use structured bindings instead of `std::tie`? Structured binding is more modern and concise I reckon.
```suggestion
        auto &&[parity, rev_wires] =
```

Reply:
`parity` and `rev_wires` are data members of the class (`parity` needs to be available to the `operator()`), and structured binding requires defining new variables. So if I use structured binding I will have to define a new variable and copy to the `parity` member, which would require an extra step?